### ISSUE TITLE
Sending a request to /api/permissions/{user}  raises an exception

### ISSUE THREAD

**Issue by Samarth-Agarwal1** [CSCfi/rems#2609]
I am using REMS and Keycloak in separate docker containers. `OIDC` is being used for authentication and I am using the test data. My config does include `:enable-permissions-api true`.

1. Start up the Keycloak Docker container.
2. Add a user named `user` in Keycloak.
3. Add a user named `owner` in Keycloak.
4. Start up the REMS Docker container.
5. In the browser, login to REMS as `owner`
6. In the REMS Docker container, give `owner` owner permissions by executing `java -Drems.config=config/config.edn -jar rems.jar grant-role owner [ID of owner as generated by Keycloak]`
7. In the browser, logout of REMS as `owner`
8. In the browser, login to REMS as `user`
9. Apply for `Auto-approve workflow`.
10. In the browser, login to REMS as `owner`
11. Go to `/swagger-ui` endpoint
12. Make a request for `/api/permissions/{user}` with the following values: `x-rems-api-key: 42`, `x-rems-user-id: [ID of owner as generated by Keycloak]`, `user: [ID of user as generated by Keycloak]`
13. The response status code is 500 Server Error with the following body:
```
{
  "type": "unknown-exception",
  "class": "java.lang.IllegalArgumentException"
}
```
The result I expect is a response status code of 200 with the body being `user`'s permissions in ga4gh visa format. `user` only applied to `Auto-approve workflow` so that should be the only permission returned.
The actual result is that an exception seems to be raised.

**Comment by Macroz** [comment]
Hi,

Indeed there seems to be an exception fetching the permissions. Could you perhaps copy the stacktrace from the logs? Otherwise it's difficult to say which part of the code caused the exception.

Also can you confirm that the GA4GH is enabled and keys configured as in [GA4GH](https://github.com/CSCfi/rems/blob/master/docs/ga4gh-visas.md)? I am able to fetch the passports with the dev settings myself.

**Comment by Samarth-Agarwal1** [comment]
```
rems_1         | 2021-03-23 16:47:06,121 [qtp1819689777-21 61aa8282-04e4-42a8-88d6-1c5c54a2f5e5 GET /api/permissions/56010b33-0532-4dfe-899c-82c41271e077] INFO  rems.middleware - > :get /api/permissions/56010b33-0532-4dfe-899c-82c41271e077 lang: :en user: {:eppn 61aa8282-04e4-42a8-88d6-1c5c54a2f5e5, :mail nil, :commonName nil} api-key roles: #{:logged-in :owner :api-key}
rems_1         | 2021-03-23 16:47:06,129 [qtp1819689777-21 61aa8282-04e4-42a8-88d6-1c5c54a2f5e5 GET /api/permissions/56010b33-0532-4dfe-899c-82c41271e077] ERROR compojure.api.exception - No method in multimethod 'jwk->private-key' for dispatch value: null
rems_1         | java.lang.IllegalArgumentException: No method in multimethod 'jwk->private-key' for dispatch value: null
rems_1         | 	at clojure.lang.MultiFn.getFn(MultiFn.java:156)
rems_1         | 	at clojure.lang.MultiFn.invoke(MultiFn.java:229)
rems_1         | 	at buddy.core.keys$jwk__GT_private_key.invokeStatic(keys.clj:73)
rems_1         | 	at buddy.core.keys$jwk__GT_private_key.invoke(keys.clj:70)
rems_1         | 	at rems.ga4gh$sign_visa.invokeStatic(ga4gh.clj:53)
rems_1         | 	at rems.ga4gh$sign_visa.invoke(ga4gh.clj:51)
rems_1         | 	at rems.ga4gh$entitlement__GT_visa.invokeStatic(ga4gh.clj:97)
rems_1         | 	at rems.ga4gh$entitlement__GT_visa.invoke(ga4gh.clj:96)
rems_1         | 	at clojure.core$mapv$fn__8445.invoke(core.clj:6912)
rems_1         | 	at clojure.core.protocols$fn__8159.invokeStatic(protocols.clj:168)
rems_1         | 	at clojure.core.protocols$fn__8159.invoke(protocols.clj:124)
rems_1         | 	at clojure.core.protocols$fn__8114$G__8109__8123.invoke(protocols.clj:19)
rems_1         | 	at clojure.core.protocols$seq_reduce.invokeStatic(protocols.clj:31)
rems_1         | 	at clojure.core.protocols$fn__8146.invokeStatic(protocols.clj:75)
rems_1         | 	at clojure.core.protocols$fn__8146.invoke(protocols.clj:75)
rems_1         | 	at clojure.core.protocols$fn__8088$G__8083__8101.invoke(protocols.clj:13)
rems_1         | 	at clojure.core$reduce.invokeStatic(core.clj:6828)
rems_1         | 	at clojure.core$mapv.invokeStatic(core.clj:6903)
rems_1         | 	at clojure.core$mapv.invoke(core.clj:6903)
rems_1         | 	at rems.ga4gh$entitlements__GT_passport.invokeStatic(ga4gh.clj:100)
rems_1         | 	at rems.ga4gh$entitlements__GT_passport.invoke(ga4gh.clj:99)
rems_1         | 	at rems.db.entitlements$get_entitlements_for_permissions_api.invokeStatic(entitlements.clj:43)
rems_1         | 	at rems.db.entitlements$get_entitlements_for_permissions_api.invoke(entitlements.clj:42)
rems_1         | 	at rems.api.permissions$fn__30696.invokeStatic(permissions.clj:38)
rems_1         | 	at rems.api.permissions$fn__30696.invoke(permissions.clj:27)
rems_1         | 	at compojure.core$wrap_response$fn__5865.invoke(core.clj:158)
rems_1         | 	at compojure.core$pre_init$fn__5918.invoke(core.clj:331)
rems_1         | 	at compojure.api.coercion$wrap_coerce_response$fn__7366.invoke(coercion.clj:89)
rems_1         | 	at compojure.core$pre_init$fn__5920$fn__5923.invoke(core.clj:338)
rems_1         | 	at compojure.core$wrap_route_middleware$fn__5849.invoke(core.clj:127)
rems_1         | 	at compojure.core$wrap_route_info$fn__5854.invoke(core.clj:137)
rems_1         | 	at compojure.core$wrap_route_matches$fn__5858.invoke(core.clj:146)
rems_1         | 	at compojure.core$wrap_routes$fn__5930.invoke(core.clj:351)
rems_1         | 	at compojure.api.routes.Route.invoke(routes.clj:89)
rems_1         | 	at compojure.core$routing$fn__5873.invoke(core.clj:185)
rems_1         | 	at clojure.core$some.invokeStatic(core.clj:2701)
rems_1         | 	at clojure.core$some.invoke(core.clj:2692)
rems_1         | 	at compojure.core$routing.invokeStatic(core.clj:185)
rems_1         | 	at compojure.core$routing.doInvoke(core.clj:182)
rems_1         | 	at clojure.lang.RestFn.applyTo(RestFn.java:139)
rems_1         | 	at clojure.core$apply.invokeStatic(core.clj:667)
rems_1         | 	at clojure.core$apply.invoke(core.clj:660)
rems_1         | 	at compojure.core$routes$fn__5877.invoke(core.clj:192)
rems_1         | 	at compojure.api.compojure_compat$make_context$handler__10810.invoke(compojure_compat.clj:26)
rems_1         | 	at compojure.api.compojure_compat$make_context$fn__10812.invoke(compojure_compat.clj:34)
rems_1         | 	at compojure.api.routes.Route.invoke(routes.clj:89)
rems_1         | 	at compojure.core$routing$fn__5873.invoke(core.clj:185)
rems_1         | 	at clojure.core$some.invokeStatic(core.clj:2701)
rems_1         | 	at clojure.core$some.invoke(core.clj:2692)
rems_1         | 	at compojure.core$routing.invokeStatic(core.clj:185)
rems_1         | 	at compojure.core$routing.doInvoke(core.clj:182)
rems_1         | 	at clojure.lang.RestFn.applyTo(RestFn.java:139)
rems_1         | 	at clojure.core$apply.invokeStatic(core.clj:667)
rems_1         | 	at clojure.core$apply.invoke(core.clj:660)
rems_1         | 	at compojure.core$routes$fn__5877.invoke(core.clj:192)
rems_1         | 	at compojure.api.routes.Route.invoke(routes.clj:89)
rems_1         | 	at compojure.core$routing$fn__5873.invoke(core.clj:185)
rems_1         | 	at clojure.core$some.invokeStatic(core.clj:2701)
rems_1         | 	at clojure.core$some.invoke(core.clj:2692)
rems_1         | 	at compojure.core$routing.invokeStatic(core.clj:185)
rems_1         | 	at compojure.core$routing.doInvoke(core.clj:182)
rems_1         | 	at clojure.lang.RestFn.applyTo(RestFn.java:139)
rems_1         | 	at clojure.core$apply.invokeStatic(core.clj:667)
rems_1         | 	at clojure.core$apply.invoke(core.clj:660)
rems_1         | 	at compojure.core$routes$fn__5877.invoke(core.clj:192)
rems_1         | 	at compojure.core$routing$fn__5873.invoke(core.clj:185)
rems_1         | 	at clojure.core$some.invokeStatic(core.clj:2701)
rems_1         | 	at clojure.core$some.invoke(core.clj:2692)
rems_1         | 	at compojure.core$routing.invokeStatic(core.clj:185)
rems_1         | 	at compojure.core$routing.doInvoke(core.clj:182)
rems_1         | 	at clojure.lang.RestFn.applyTo(RestFn.java:139)
rems_1         | 	at clojure.core$apply.invokeStatic(core.clj:667)
rems_1         | 	at clojure.core$apply.invoke(core.clj:660)
rems_1         | 	at compojure.core$routes$fn__5877.invoke(core.clj:192)
rems_1         | 	at compojure.core$make_context$handler__5905.invoke(core.clj:286)
rems_1         | 	at compojure.core$make_context$fn__5909.invoke(core.clj:296)
rems_1         | 	at compojure.api.routes.Route.invoke(routes.clj:89)
rems_1         | 	at compojure.core$routing$fn__5873.invoke(core.clj:185)
rems_1         | 	at clojure.core$some.invokeStatic(core.clj:2701)
rems_1         | 	at clojure.core$some.invoke(core.clj:2692)
rems_1         | 	at compojure.core$routing.invokeStatic(core.clj:185)
rems_1         | 	at compojure.core$routing.doInvoke(core.clj:182)
rems_1         | 	at clojure.lang.RestFn.applyTo(RestFn.java:139)
rems_1         | 	at clojure.core$apply.invokeStatic(core.clj:667)
rems_1         | 	at clojure.core$apply.invoke(core.clj:660)
rems_1         | 	at compojure.core$routes$fn__5877.invoke(core.clj:192)
rems_1         | 	at compojure.api.routes.Route.invoke(routes.clj:89)
rems_1         | 	at ring.swagger.middleware$wrap_swagger_data$fn__11787.invoke(middleware.clj:35)
rems_1         | 	at rems.api$transaction_middleware$fn__33102$fn__33107.invoke(api.clj:128)
rems_1         | 	at clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:789)
rems_1         | 	at clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:759)
rems_1         | 	at clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:824)
rems_1         | 	at clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:759)
rems_1         | 	at rems.api$transaction_middleware$fn__33102.invoke(api.clj:126)
rems_1         | 	at ring.middleware.cors$handle_cors.invokeStatic(cors.cljc:178)
rems_1         | 	at ring.middleware.cors$handle_cors.invoke(cors.cljc:167)
rems_1         | 	at ring.middleware.cors$wrap_cors$fn__33064.invoke(cors.cljc:205)
rems_1         | 	at compojure.api.middleware$wrap_swagger_data$fn__9267.invoke(middleware.clj:190)
rems_1         | 	at compojure.api.middleware$wrap_inject_data$fn__9245.invoke(middleware.clj:96)
rems_1         | 	at muuntaja.middleware$wrap_params$fn__8366.invoke(middleware.clj:52)
rems_1         | 	at compojure.api.middleware$wrap_exceptions$fn__9234.invoke(middleware.clj:62)
rems_1         | 	at muuntaja.middleware$wrap_format_request$fn__8378.invoke(middleware.clj:114)
rems_1         | 	at compojure.api.middleware$wrap_exceptions$fn__9234.invoke(middleware.clj:62)
rems_1         | 	at muuntaja.middleware$wrap_format_response$fn__8382.invoke(middleware.clj:132)
rems_1         | 	at muuntaja.middleware$wrap_format_negotiate$fn__8375.invoke(middleware.clj:96)
rems_1         | 	at ring.middleware.keyword_params$wrap_keyword_params$fn__7404.invoke(keyword_params.clj:53)
rems_1         | 	at ring.middleware.nested_params$wrap_nested_params$fn__7454.invoke(nested_params.clj:89)
rems_1         | 	at ring.middleware.params$wrap_params$fn__7502.invoke(params.clj:67)
rems_1         | 	at compojure.api.middleware$wrap_inject_data$fn__9245.invoke(middleware.clj:96)
rems_1         | 	at compojure.api.routes.Route.invoke(routes.clj:89)
rems_1         | 	at rems.api$audit_log_middleware$fn__33096.invoke(api.clj:112)
rems_1         | 	at clojure.lang.Var.invoke(Var.java:384)
rems_1         | 	at compojure.core$routing$fn__5873.invoke(core.clj:185)
rems_1         | 	at clojure.core$some.invokeStatic(core.clj:2701)
rems_1         | 	at clojure.core$some.invoke(core.clj:2692)
rems_1         | 	at compojure.core$routing.invokeStatic(core.clj:185)
rems_1         | 	at compojure.core$routing.doInvoke(core.clj:182)
rems_1         | 	at clojure.lang.RestFn.applyTo(RestFn.java:139)
rems_1         | 	at clojure.core$apply.invokeStatic(core.clj:667)
rems_1         | 	at clojure.core$apply.invoke(core.clj:660)
rems_1         | 	at compojure.core$routes$fn__5877.invoke(core.clj:192)
rems_1         | 	at rems.middleware$wrap_fix_location_header$fn__32910.invoke(middleware.clj:190)
rems_1         | 	at rems.middleware$wrap_unauthorized_and_forbidden$fn__32877.invoke(middleware.clj:139)
rems_1         | 	at rems.middleware$wrap_logging$fn__32880.invoke(middleware.clj:161)
rems_1         | 	at rems.middleware$wrap_i18n$fn__32871.invoke(middleware.clj:122)
rems_1         | 	at rems.middleware$wrap_role_headers$fn__32829.invoke(middleware.clj:84)
rems_1         | 	at rems.middleware$wrap_context$fn__32824$fn__32825.invoke(middleware.clj:80)
rems_1         | 	at rems.logging$with_mdc_STAR_.invokeStatic(logging.clj:21)
rems_1         | 	at rems.logging$with_mdc_STAR_.invoke(logging.clj:9)
rems_1         | 	at rems.middleware$wrap_context$fn__32824.invoke(middleware.clj:79)
rems_1         | 	at rems.middleware$wrap_user$fn__32819$fn__32820.invoke(middleware.clj:67)
rems_1         | 	at rems.logging$with_mdc_STAR_.invokeStatic(logging.clj:21)
rems_1         | 	at rems.logging$with_mdc_STAR_.invoke(logging.clj:9)
rems_1         | 	at rems.middleware$wrap_user$fn__32819.invoke(middleware.clj:66)
rems_1         | 	at rems.middleware$wrap_api_key_or_csrf_token$fn__32816.invoke(middleware.clj:56)
rems_1         | 	at buddy.auth.middleware$wrap_authentication$fn__28596.invoke(middleware.clj:59)
rems_1         | 	at rems.auth.auth$wrap_uses_valid_api_key$fn__28933.invoke(auth.clj:36)
rems_1         | 	at ring.middleware.flash$wrap_flash$fn__31600.invoke(flash.clj:39)
rems_1         | 	at ring.middleware.session$wrap_session$fn__31809.invoke(session.clj:108)
rems_1         | 	at ring.middleware.keyword_params$wrap_keyword_params$fn__7404.invoke(keyword_params.clj:53)
rems_1         | 	at ring.middleware.nested_params$wrap_nested_params$fn__7454.invoke(nested_params.clj:89)
rems_1         | 	at ring.middleware.multipart_params$wrap_multipart_params$fn__29835.invoke(multipart_params.clj:171)
rems_1         | 	at ring.middleware.params$wrap_params$fn__7502.invoke(params.clj:67)
rems_1         | 	at ring.middleware.cookies$wrap_cookies$fn__31772.invoke(cookies.clj:214)
rems_1         | 	at ring.middleware.absolute_redirects$wrap_absolute_redirects$fn__31880.invoke(absolute_redirects.clj:47)
rems_1         | 	at ring.middleware.content_type$wrap_content_type$fn__12241.invoke(content_type.clj:34)
rems_1         | 	at ring.middleware.default_charset$wrap_default_charset$fn__31860.invoke(default_charset.clj:31)
rems_1         | 	at ring.middleware.not_modified$wrap_not_modified$fn__12266.invoke(not_modified.clj:61)
rems_1         | 	at ring.middleware.x_headers$wrap_x_header$fn__31571.invoke(x_headers.clj:22)
rems_1         | 	at ring.middleware.x_headers$wrap_x_header$fn__31571.invoke(x_headers.clj:22)
rems_1         | 	at ring.middleware.x_headers$wrap_x_header$fn__31571.invoke(x_headers.clj:22)
rems_1         | 	at rems.middleware$wrap_cache_control$fn__32936.invoke(middleware.clj:208)
rems_1         | 	at rems.middleware$wrap_internal_error$fn__32858.invoke(middleware.clj:107)
rems_1         | 	at rems.middleware$wrap_request_context$fn__32889$fn__32890.invoke(middleware.clj:170)
rems_1         | 	at rems.logging$with_mdc_STAR_.invokeStatic(logging.clj:21)
rems_1         | 	at rems.logging$with_mdc_STAR_.invoke(logging.clj:9)
rems_1         | 	at rems.middleware$wrap_request_context$fn__32889.invoke(middleware.clj:168)
rems_1         | 	at ring.adapter.jetty$proxy_handler$fn__37560.invoke(jetty.clj:27)
rems_1         | 	at ring.adapter.jetty.proxy$org.eclipse.jetty.server.handler.AbstractHandler$ff19274a.handle(Unknown Source)
rems_1         | 	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
rems_1         | 	at org.eclipse.jetty.server.Server.handle(Server.java:501)
rems_1         | 	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:383)
rems_1         | 	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:556)
rems_1         | 	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:375)
rems_1         | 	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:273)
rems_1         | 	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
rems_1         | 	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
rems_1         | 	at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
rems_1         | 	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336)
rems_1         | 	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313)
rems_1         | 	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
rems_1         | 	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129)
rems_1         | 	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:375)
rems_1         | 	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:806)
rems_1         | 	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:938)
rems_1         | 	at java.base/java.lang.Thread.run(Unknown Source)
rems_1         | 2021-03-23 16:47:06,131 [qtp1819689777-21 61aa8282-04e4-42a8-88d6-1c5c54a2f5e5 GET /api/permissions/56010b33-0532-4dfe-899c-82c41271e077] INFO  rems.middleware - < :get /api/permissions/56010b33-0532-4dfe-899c-82c41271e077 500 
```

Hi @Macroz 

The above code snippet is the stacktrace from the logs. 

I have the following line in my config:
```
:enable-permissions-api true
```

I have not set `:ga4gh-visa-private-key` and `:ga4gh-visa-public-key` in my config. I will set these two variables and get back to you.

**Comment by Samarth-Agarwal1** [comment]
So I have set `:ga4gh-visa-private-key` and `:ga4gh-visa-public-key` and now when I send a request I get back a response with 200 status code and body in a format such as:
```
{
  "ga4gh_passport_v1": [
    <JSON Web token of passport>
  ]
}
```

**Comment by opqdonut** [comment]
@Samarth-Agarwal1 great! We'll improve the error messages a bit, but meanwhile I think this ticket can be closed?

**Comment by Macroz** [comment]
I have created a new ticket about the improvements to validation here https://github.com/CSCfi/rems/issues/2613

**Comment by Samarth-Agarwal1** [comment]
@opqdonut Is there a way for REMS to return the passport without it being in JSON web token format? When I first started using the REMS Docker container in early February, I remember that I just had to set `:enable-permissions-api true` in my config and sending a request to `/api/permissions/{user}` would return a 200 status code with the body being in the format of:
```
{
  "ga4gh_passport_v1": [
    <Passport in the format of a JSON dictionary>
  ]
}
```

And randomly, about two weeks ago I started getting that exception I posted.

Was there a recent change where public/private JSON keys have to be provided and the return body became a JWT?

**Comment by Macroz** [comment]
I think the [standard specifies the use of JWT tokens](https://github.com/ga4gh-duri/ga4gh-duri.github.io/blob/master/researcher_ids/ga4gh_passport_v1.md#passport-visa).

The only change from this year has been to have the timestamp in seconds, not milliseconds. All the features have otherwise remained the same.

May I ask what are you using this API for exactly? Maybe we can help.
- If you want to use GA4GH specifically, I think you have to use it as specified with the tokens. Having proper signing keys is in the standard.
- One can also read the claims of the JWT token without validating it, but that defeats the purpose of JWT. This could be used for some automated tests however. The process is roughly: split by `.`, base64 decode each part you are interested in, then parse the strings to JSON.
- If you are just interested in the entitlements we have also the `/api/entitlements` which may be what you'd like to use.

**Comment by opqdonut** [comment]
We reworked the experimental permissions API to produce signed JWTs in PRs #2344 #2384 (merged September 2020, released in 2020-12-09 in v2.14). At the same time REMS started requiring a signing key for the JWTs to be configured. This could be highlighted more clearly in the 2.14 changelog, sorry for that! However, the API _is_ experimental, so be prepared for changes.

The permissions API is trying to implement the GA4GH spec, which is itself a work in progress. You can find some more information and links in our docs: https://github.com/CSCfi/rems/blob/master/docs/ga4gh-visas.md

**Comment by Samarth-Agarwal1** [comment]
Hi,

This ticket can be closed now. I have updated my code to handle JWT's being returned from `/api/permissions/{user}`.

Thanks for the help!
